[ADD] Make snippets have a max number of views

NoeMastrorillo opened the pull request:
```
The user is now able to set a maximum number of views/uses on his
snippets. By default, this number is infinite. The user can turn off the
infinite views by unchecking the 'infinite_views checkbox and by
entering a views number in the 'max_views' input field (it has to be a
number greater than 0). Every time the snippet's page is visited, its
views left decrease, and when it gets to 0, the snippet is deleted from
the database.
```

NoeMastrorillo committed:
add: Make snippets have a max number of views

The user is now able to set a maximum number of views/uses on his
snippets. By default, this number is infinite. The user can turn off the
infinite views by unchecking the 'infinite_views checkbox and by
entering a views number in the 'max_views' input field (it has to be a
number greater than 0). Every time the snippet's page is visited, its
views left decrease, and when it gets to 0, the snippet is deleted from
the database.

Related to #5

Julien00859 commented on bin/controller.py:
```diff
@@ -38,22 +39,34 @@ def post_new():
             ext = "."+langtoext.get(lang, lang)
 
-    snippet = Snippet.create(code.decode('utf-8'))
+    max_views = math.inf
```
float('+inf'), l'import à `math` n'est plus nécessaire

Julien00859 commented on bin/controller.py:
```diff
@@ -38,22 +39,34 @@ def post_new():
+        try:
+            max_views = int(forms.get('max_views'))
+            if max_views < 1:
+                raise bt.HTTPError(422, 'Max views should be greater than 0')                   # TODO: error handling in front-side
```
Le code d'erreur 422 est une extension WebDAV, on ne fait pas de WebDAV ici. Il faut utiliser 400.

Julien00859 commented on bin/controller.py:
```diff
@@ -38,22 +39,34 @@ def post_new():
+            if max_views < 1:
+                raise bt.HTTPError(422, 'Max views should be greater than 0')                   # TODO: error handling in front-side
+        except ValueError as e:
+            raise bt.HTTPError(422, f"Max views '{forms.get('max_views')}' is not a number")    # TODO: error handling in front-side
```
HTTP Code 400

Julien00859 commented on bin/models.py:
```diff
@@ -12,28 +13,44 @@
+            if views_left <= 0:
+                database.delete(snippet_id)
+            else:
+                database.hset(snippet_id, 'views_left', int(views_left))
```
Use `HINCRBY` https://redis.io/commands/hincrby

Julien00859 commented on bin/models.py:
```diff
@@ -12,28 +13,44 @@
-        def get(cls, ident):
-            return cls.db.get(ident)
+        def hgetall(cls, ident):
+            return cls.db[ident]
```
Pour rester consistant avec redis, `cls.db.get`

Julien00859 commented on bin/models.py:
```diff
@@ -3,6 +3,7 @@
 from redis import Redis
 from genpw import pronounceable_passwd
 from bin import config
+import math
```
L'import n'est pas utilisé

Julien00859 commented on bin/controller.py:
```diff
@@ -38,22 +39,34 @@ def post_new():
+    if files:
+        code = next(files.values()).file.read(config.SNIPPET_MAX_SIZE)
+    elif forms:
+        code = forms.get('code', '').encode('latin-1')
```
Attention auteurs de #26 ce changement va rentrer en conflit.

Julien00859 reviewed (changes requested):
L'historique est foireux... Très moche commit de merge...

Pour réparer je suggère de faire une nouvelle branche sur base de main et de `git cherry-pick` le commit e7676bf qui semble être le seul qui ajoute du contenu et de régler les potentiels conflits. Si le résultat final est celui escompté alors tu peux renommer cette nouvelle branche comme 5-snippet-views-limit-noe en supprimant l'originale et force-pusher avec lease.

fusetim commented on bin/controller.py:
```diff
@@ -38,22 +39,34 @@ def post_new():
+    max_views = math.inf
+    if not forms.get('infinite_views'):
+        try:
+            max_views = int(forms.get('max_views'))
```
Il serait bien d'utiliser une valeur par défaut ici (comme +infini)
**Suggestion éditée:**
```suggestion
            max_views = float(forms.get('max_views', '+inf'))
```
et éventuellement transformer `infinite_views` en `max_views=+inf`

Julien00859 replied:
Il faut utiliser `float`, `+inf` n'est pas un entier valide

fusetim replied:
Malheureusement `forms.get("max_view")` peut renvoyer None; et donc max_view sera None

NoeMastrorillo replied:
Ah j'ai oublié de changer ça...

NoeMastrorillo committed:
fix: Add a default value to max_views

forms.get('max_views') could return None and so
int(forms.get('max_views')) could return an error. By adding a default
value with forms.get('max_views', float('+inf')), it will prevent this
error to occur.

NoeMastrorillo committed:
imp: Improve arguments algorithm and remove todos

The server now takes into account request's multipart file and form
posts with new default values for parameters, and errors handling with
`contextlib.suppress()` has been implemented instead of basics
try-except statements.

Todo statements has been removed as well as the code that they were
related to, like the form for snippets maxusage for which a new issue
will be created.

The in-memory database has been replaced by an error raising because it
was not convenient at all. It will be re-implemented in a better way
later for unit-testing.

Related to #5
